src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +79,181 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
+        return;
 
-        using (Logger.LogBlock(FunctionId.EventHookup_Generate_Handler, cancellationToken))
+        async Task ExecuteCommandWorkerAsync()
         {
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
+            var textView = args.TextView;
+            var subjectBuffer = args.SubjectBuffer;
+
+            var factory = document.Project.Solution.Workspace.Services.GetRequiredService<IBackgroundWorkIndicatorFactory>();
+            using var waitContext = factory.Create(
+                textView,
+                applicableToSpan,
+                CSharpEditorResources.Generating_event);
+
+            var cancellationToken = waitContext.UserCancellationToken;
 
-            var workspace = textView.TextSnapshot.TextBuffer.GetWorkspace();
-            if (workspace == null)
+            var eventHandlerMethodName = await getEventNameTask.WithCancellation(cancellationToken).ConfigureAwait(false);
+
+            var solutionAndRenameSpan = await TryGetNewSolutionWithAddedMethodAsync(
+                document, eventHandlerMethodName, initialCaretPoint.Position, cancellationToken).ConfigureAwait(false);
+
+            // We're about to make an edit ourselves.  so disable the cancellation that happens on editing.
+            waitContext.CancelOnEdit = false;
+
+            // switch back to the UI thread.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            // If anything changed in the view between computation and application, bail out.
+            if (solutionAndRenameSpan is null ||
+                applicableToSpan.Snapshot.Version != subjectBuffer.CurrentSnapshot.Version ||
+                textView.GetCaretPoint(subjectBuffer) != initialCaretPoint)
             {
                 nextHandler();
-                EventHookupSessionManager.CancelAndDismissExistingSessions();
                 return;
             }
 
-            var document = textView.TextSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-            Contract.ThrowIfNull(document, "Event Hookup could not find the document for the IBufferView.");
-
-            var position = textView.GetCaretPoint(subjectBuffer).Value.Position;
-            var solutionWithEventHandler = CreateSolutionWithEventHandler(
-                document,
-                eventHandlerMethodName,
-                position,
-                out var plusEqualTokenEndPosition,
-                _globalOptions,
-                cancellationToken);
-
-            Contract.ThrowIfNull(solutionWithEventHandler, "Event Hookup could not create solution with event handler.");
-
-            // The new solution is created, so start user observable changes
+            var (solutionWithEventHandler, renameSpan) = solutionAndRenameSpan.Value;
+            var workspace = document.Project.Solution.Workspace;
```

Comment:
application to the workspace and start of rename session both happen on the UI thread as required.